### PR TITLE
More options for userscript matches

### LINKED ISSUE #23: More options for userscript matches (not only URLs)
Let's say a userscript only works on user profiles. The current match would currently be `https://scratch.mit.edu/users/*/`, why not just `userprofile`? And other potential filters for common use.

### PR DESCRIPTION
**Pull request by TheColaber**
70 changed files ahh

Resolve #23

Please discuss with me what I should add/remove and if I have messed up any addons since I'm sleepy. I'm more specifically focused on dark-www and scratchr3.

### CODE REVIEW

**Comment by apple502j** [comment]
think we can do it better

**Comment by mxmou** [comment]
I don't think everything needs to have named matches - for most pages matching by URL is sufficient. The following would probably be helpful:
* `project`: matches `/projects/*/`, `/projects/*/editor` and `/projects/*/fullscreen` but not `/projects/*/remixes` or `/projects/*/studios`
* `forums`: matches `/discuss/*` but not `/discuss/m/*`
* `403`, `404`, `500`: error pages (other matches, including URL patterns, should never match error pages)
* `scratchr2` and `scrathchwww`: matches by page type
* `studios_scratchr2` and `studios_scratchwww`: see https://github.com/ScratchAddons/ScratchAddons/issues/1972#issuecomment-808743331

**Comment by RedGuy12** [comment]
You’re not in the discord…we’re also going to use theese for settings page filtering in the future.

**Comment by apple502j** [comment]
Closing; I will implement better version this weekend

**Comment by Hans5958** [comment]
...and this is why feature discussions should take on GitHub, not Discord.

**Comment by mxmou** [comment]
> You’re not in the discord…we’re also going to use theese for settings page filtering in the future.

I don't think this is a good idea. Addons should be able to define how they appear on the settings page separately from their functionality.

**Comment by TheColaber** [comment]
> I don't think this is a good idea. Addons should be able to define how they appear on the settings page separately from their functionality.

Hmm, but if you have matches for the forums, why would you also add a tag when the settings page could just look for the forum matches?

**Comment by mxmou** [comment]
> Hmm, but if you have matches for the forums, why would you also add a tag when the settings page could just look for the forum matches?

One reason is that there are addons that don't set their match to the forums but are closely related to them (discuss button). There are also addons that have forum matches to implement forum-specific functionality but aren't otherwise forum addons (infinite scrolling).

Another reason is that if we add subcategories, editor addons will need to set them explicitly anyway. Why would website addons work differently?

**Comment by Hans5958** [comment]
Should we move to #23 for the time being?

**Comment by TheColaber** [comment]
> Closing; I will implement better version this weekend

I think you missed your chance. Reopening till you actually open one.

File: background/get-userscripts.js [hunk 1]
```diff
@@ -262,6 +262,29 @@ function userscriptMatches(data, scriptOrStyle, addonId) {
...
+    case "profiles":
+      pattern = "https://scratch.mit.edu/users/*/";
+      break;
+    case "studio":
+      pattern = "https://scratch.mit.edu/studios/*";
+      break;
+    case "forums":
+      pattern = "https://scratch.mit.edu/studios/*";
+      break;
+    case "forums_topic":
```
Comment (GrahamSH-LLK) [inline background/get-userscripts.js:281]: Shouldn't these be camelCase?

Comment (TheColaber) [reply background/get-userscripts.js:281]: For what reason?

Comment (mxmou) [reply background/get-userscripts.js:281]: @TheColaber Everything in `addon.json` is currently camel case.

**Comment by Hans5958** [comment]
Why should we do this? I think we shouldn't make matches difficult. Having those URL is already simple, but it seems we're going to make this a little be abstract. Why can't we have simple things and keep it as it is?

**Comment by mxmou** [comment]
> Why should we do this? I think we shouldn't make matches difficult. Having those URL is already simple, but it seems we're going to make this to be a little abstract. Why can't we have simple things and keep it as it is?

I think it's best to only add an alternative where it would be more convenient than matching by URL - https://github.com/ScratchAddons/ScratchAddons/pull/2257#issuecomment-824921416

**Comment by WorldLanguages** [comment]
Closed in favor of #2305

**Pull request closed without merging**